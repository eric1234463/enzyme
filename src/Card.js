import React, { Component } from 'react';

class Card extends Component {
  constructor() {
    super();
    this.state = {
      isShow: false,
    };
  }

  handleOnClick = () => {
    this.setState({
      isShow: !this.state.isShow,
    });
  };

  render() {
    const { rows, tag } = this.props;
    const { isShow } = this.state;
    return (
      <div id="container" onClick={this.handleOnClick}>
        {rows.length === 0 && <div className="null-message">Null</div>}
        {rows.map((row, index) => {
          if (row.status === 'active') {
            return (
              <div className="label" key={index}>
                {row.name}
              </div>
            );
          }
          return null;
        })}

        {tag === 'DONE' && (
          <div id="status-tag" className="done-tag">
            Done
          </div>
        )}
        {tag === 'PREPARE' && (
          <div id="status-tag" className="prepare-tag">
            Prepare
          </div>
        )}

        {isShow && <div>Show Label</div>}
      </div>
    );
  }
}

export default Card;
