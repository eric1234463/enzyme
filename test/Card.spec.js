import React from 'react';
import { shallow, render, mount } from 'enzyme';
import Card from '../src/Card';

describe('Card', () => {
  const rows = [
    {
      name: '123',
      status: 'active'
    },
    {
      name: '123',
      status: 'inactive'
    }
  ];

  it('should render correctly as snapshot', () => {
    const mockOnClick = jest.fn();
    Card.prototype.handleOnClick = mockOnClick;

    const wrapper = shallow(<Card rows={rows} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render 1 active div with class label', () => {
    const activeRows = [
      {
        name: '123',
        status: 'active'
      },
      {
        name: '123',
        status: 'active'
      }
    ];
    const wrapper = render(<Card rows={activeRows} />);

    expect(wrapper.find('div.label')).toHaveLength(2);
  });

  it('should render 0 active div with class label', () => {
    const activeRows = [
      {
        name: '123',
        status: 'inactive'
      },
      {
        name: '123',
        status: 'inactive'
      }
    ];
    const wrapper = render(<Card rows={activeRows} />);

    expect(wrapper.find('div.label')).toHaveLength(0);
  });

  it('should render done status tag', () => {
    const activeRows = [
      {
        name: '123',
        status: 'inactive'
      },
      {
        name: '123',
        status: 'inactive'
      }
    ];
    const wrapper = render(<Card rows={activeRows} tag={'DONE'} />);

    expect(wrapper.find('div.label')).toHaveLength(0);

    expect(wrapper.find('#status-tag')).toHaveLength(1);

    expect(wrapper.find('#status-tag').html()).toEqual('Done');

    expect(wrapper.find('#status-tag').hasClass('done-tag')).toEqual(true);
  });

  it('should contain null message', () => {
    const wrapper = mount(<Card rows={[]} />);

    expect(wrapper.contains(<div className='null-message'>Null</div>)).toEqual(true);

    expect(wrapper.find('.div.label').exists()).toEqual(false);

  });

  it('when ComponentDidMount should call addEventListener && unmount should call removeEventListener', () => {
    const adder = jest.spyOn(global, 'addEventListener').mockImplementation(() => {});

    const remover = jest.spyOn(global, 'removeEventListener').mockImplementation(() => {});
    const wrapper = mount(<Card rows={[]} />);
    
    expect(adder).toBeCalled();
    
    wrapper.unmount();
    
    expect(remover).toBeCalled();
  });

  it('when container click should change state isShow', () => {
    const wrapper = mount(<Card rows={[]} />);

    // Simulate OnClick Event
    wrapper.find('#container').simulate('click');

    expect(wrapper).toMatchSnapshot();

    expect(wrapper.state().isShow).toEqual(true);
  })
});
