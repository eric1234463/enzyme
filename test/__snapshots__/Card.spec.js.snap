// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should contain null message 1`] = `
<Card
  rows={Array []}
>
  <div
    id="container"
    onClick={[Function]}
  >
    <div
      className="null-message"
    >
      Null
    </div>
    <div>
      Show Label
    </div>
  </div>
</Card>
`;

exports[`Card should render correctly as snapshot 1`] = `
<div
  id="container"
  onClick={[Function]}
>
  <div
    className="label"
    key="0"
  >
    123
  </div>
</div>
`;
